include/ginkgo/core/matrix/dense.hpp

```diff
@@ -456,26 +491,6 @@ class Dense
      */
     std::unique_ptr<Dense> row_gather(const array<int64>* gather_indices) const;
 
-    /**
-     * Copies the given rows from this matrix into `row_collection`
-     *
-     * @param gather_indices  pointer to an array containing row indices
-     *                        from this matrix. It may contain duplicates.
-     * @param row_collection  pointer to a Dense matrix that will store the
-     *                      gathered rows:
-     *                      `row_collection(i,j) = input(gather_indices(i), j)`
-     *                      It must have the same number of columns as this
-     *                      matrix and `gather_indices->get_num_elems()` rows.
-     */
-    void row_gather(const array<int32>* gather_indices,
-                    Dense* row_collection) const;
```

Comment:
where be this moved to?

Reply:
pointer_param<LinOp> accepts Dense*, so this overload is no longer necessary

Reply:
I think the original one also accepts Dense*, but it can avoid casting to LinOp* and recasting back the Dense?

Reply:
Basically, we can't have two overloads that take `pointer_param<Dense>` and `pointer_param<LinOp>`, because the call would then be ambiguous for any `Dense` pointer we pass in. The reason for this is that with raw pointers, more generic types (LinOp) are a worse fit than more specific types (Dense), but this is not true if we have converting constructors like the ones for `pointer_param(T*)`

Reply:
With `pointer_param`, we will only allow the BaseType as the public interface, and then we need to dynamic_cast to the accepted types.

Reply:
That is true, though dynamic_cast hasn't really been a performance issue so far. I think in the long term (Ginkgo 2.0), we should aim for making the parameter a `Vector` base class, instead of the current overloads being either too specific `Dense<double>` or too generic `LinOp`

Reply:
after checking https://godbolt.org/z/8bcbYK8G4 again, the shared_ptr does not work, too. It only work when have derived and base function signature

Reply:
yes, we always need exact matches, otherwise the ambiguity will come into play